File: src/main/java/com/larryhsiao/clotho/file/ToFile.java
package com.larryhsiao.clotho.file;

import com.larryhsiao.clotho.Action;
import com.larryhsiao.clotho.io.ProgressedCopy;

import java.io.*;
import java.util.function.Function;

/**
 * Copy given stream ot file.
 */
public class ToFile implements Action {
    private final int bufferSize;
    private final InputStream input;
    private final File dst;
    private final Function<Integer, Void> progress;

    public ToFile(
        InputStream input,
        File dst,
        Function<Integer, Void> progress,
        int bufferSize
    ) {
        this.bufferSize = bufferSize;
        this.input = input;
        this.dst = dst;
        this.progress = progress;
    }

    public ToFile(
        InputStream input,
        File dst,
        Function<Integer, Void> progress
    ) {
        this(input, dst, progress, 1024 * 1024 * 4);
    }


    @Override
    public void fire() {
        try {
            final OutputStream output = new FileOutputStream(dst);
            new ProgressedCopy(
                input,
                output,
                bufferSize,
                progress
            ).fire();
            output.close();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/java/com/larryhsiao/clotho/file/ToFile.kt
package com.larryhsiao.clotho.file

import com.larryhsiao.clotho.Action
import com.larryhsiao.clotho.io.ProgressedCopy
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream

/**
 * Copy source to target [File].
 */
class ToFile(
    private val input: InputStream,
    private val dst: File,
    private val progress: (copied: Int) -> Unit
) : Action {

    /**
     * File to destination file
     */
    constructor(
        src: File,
        dst: File,
        progress: (copied: Int) -> Unit
    ) : this(
        FileInputStream(src),
        dst,
        progress
    )

    override fun fire() {
        ProgressedCopy(
            input,
            FileOutputStream(dst)
        ) { progress(it) }.value()
    }
}